core/src/main/java/org/elasticsearch/common/io/Streams.java

```diff
@@ -236,4 +241,57 @@ public static void readAllLines(InputStream input, Callback<String> callback) th
             }
         }
     }
+
+    /**
+     * Wraps the given {@link BytesStreamOutput} in a {@link StreamOutput} that simply flushes when
+     * close is called. The returned values from this method also implements the {@link BytesStream}
+     * interface.
+     */
+    public static StreamOutput flushOnCloseStream(BytesStreamOutput os) {
+        return new FlushOnCloseOutputStream(os);
+    }
+
+    /**
+     * A wrapper around a {@link BytesStreamOutput} that makes the close operation a flush. This is
+     * needed as sometimes a stream will be closed but the bytes that the stream holds still need
+     * to be used and the stream cannot be closed until the bytes have been consumed.
+     */
+    private static class FlushOnCloseOutputStream extends StreamOutput implements BytesStream {
```

Comment:
I pushed https://github.com/elastic/elasticsearch/pull/23572/commits/53851fddcb96e4c8e28bbca15cfd0c6f9593095c which does the following. `BytesStream` is no longer an interface but is now an abstract class that extends `StreamOutput` and `BytesStreamOutput` extends `BytesStream`. `ReleasableBytesStream` has been removed as there is only a single implementation and the interface was only used in `TcpTransport`. The `FlushOnCloseOutputStream` now extends `BytesStream`. The new Streams method takes a `BytesStream` and returns a `BytesStream`. I think in the future we may want to consider removing `BytesStream` altogether and move this method to `StreamOutput`.